### PR TITLE
Fix bug in Path.join. Fix #4252.

### LINKED ISSUE #4252: Path.join(["/", "", "foo"]) != Path.join(["/", "foo"])
This seems buggy:

```elixir
iex(2)> ["/", "foo"] |> Path.join       
"/foo"
iex(3)> ["/", "", "foo"] |> Path.join
"foo"
```

Ruby:
```ruby
irb(main):001:0> File.join("/", "", "foo")
=> "/foo"
```

### PR DESCRIPTION
**Pull request by tuvistavie**
This should fix #4252.

### CODE REVIEW

File: lib/elixir/lib/path.ex [hunk 1]
```diff
@@ -466,7 +466,10 @@ defmodule Path do
   def join(left, right) do
     left    = IO.chardata_to_string(left)
     os_type = major_os_type()
-    do_join(left, right, os_type) |> remove_dirsep(os_type)
+    path = do_join(left, right, os_type)
+    if byte_size(path) > 1,
```
Comment (josevalim) [inline lib/elixir/lib/path.ex]: I am not very confident with this byte_size call. Maybe we could fix this by moving remove_dirsep inside the `do_join` calls below and only call it when appropriate? Can you give this alternative solution a try to see if things are better or worse? Thank you!

Comment (tuvistavie) [inline lib/elixir/lib/path.ex]: Sure, I'll give it a try!

**Comment by josevalim** [comment]
Thank you! I have added a comment!

**Comment by tuvistavie** [comment]
@josevalim I tried to move the logic to `do_join` but I am not sure if there is
a way to avoid checking the byte size, as the `/` case is a little special.

Maybe it would be better to move the last `remove_dirsep` call in a `normalize`
function or something like this and handle this logic there.

File: lib/elixir/lib/path.ex [hunk 2]
```diff
@@ -466,12 +466,14 @@ defmodule Path do
...
     os_type = major_os_type()
-    do_join(left, right, os_type) |> remove_dirsep(os_type)
+    do_join(left, right, os_type)
   end
 
-  defp do_join("", right, os_type),   do: relative(right, os_type)
-  defp do_join(left, "", _os_type),   do: left
-  defp do_join(left, right, os_type), do: remove_dirsep(left, os_type) <> "/" <> relative(right, os_type)
+  defp do_join("", right, os_type),   do: remove_dirsep(relative(right, os_type), os_type)
+  defp do_join(<<_, "">> = left, "", _os_type), do: left
```
Comment (lexmag) [inline lib/elixir/lib/path.ex]: Why do we need `""` inside the pattern?
`<<_, "">>` is the same as `<<_>>`.

**Comment by josevalim** [comment]
Maybe we can change remove_dirsep to not do anything if it receives only
"/" then?

On Tuesday, February 2, 2016, Daniel Perez <notifications@github.com> wrote:

> @josevalim <https://github.com/josevalim> I tried to move the logic to
> do_join but I am not sure if there is
> a way to avoid checking the byte size, as the / case is a little special.
>
> Maybe it would be better to move the last remove_dirsep call in a
> normalize
> function or something like this and handle this logic there.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/elixir-lang/elixir/pull/4253#issuecomment-178455464>.
>


-- 


*José Valimwww.plataformatec.com.br
<http://www.plataformatec.com.br/>Founder and Director of R&D*

**Comment by josevalim** [comment]
@tuvistavie I have added a new comment so maybe we explicitly check for `"/"` in remove_dirsep. That should achieve the same but a bit more cleanly? Thanks for trying all of those different approaches!

File: lib/elixir/lib/path.ex (hunk 2 above)
Comment (tuvistavie) [inline lib/elixir/lib/path.ex]: Thank you, I changed the code a little!

**Comment by tuvistavie** [comment]
@josevalim Sorry for the delay.

I changed the `remove_dirsep` and `do_join`
to make this a little cleaner. This should work prefectly for UNIX/Linux based OS,
but I am not very sure if we need to do something special for Windows here.

**Pull request merged**